nucliadb_utils/src/nucliadb_utils/aiopynecone/client.py

```diff
@@ -47,37 +90,320 @@ async def create_index(self, name: str, dimension: int) -> str:
+        """
+        headers = {"Api-Key": self.api_key}
+        payload = UpsertRequest(vectors=vectors)
+        post_kwargs: dict[str, Any] = {
+            "headers": headers,
+            "json": payload.model_dump(),
+        }
+        if timeout is not None:
+            post_kwargs["timeout"] = timeout
+        response = await self.http_session.post("/vectors/upsert", **post_kwargs)
+        raise_for_status(response)
+
+    def _estimate_upsert_batch_size(self, vectors: list[Vector]) -> int:
+        """
+        Estimate a batch size so that the upsert payload does not exceed the hard limit.
+        https://docs.pinecone.io/reference/quotas-and-limits#hard-limits
+        """
+        if self._upsert_batch_size is not None:
+            # Return the cached value.
+            return self._upsert_batch_size
+        # Take the dimension of the first vector as the vector dimension.
+        # Assumes all vectors have the same dimension.
+        vector_dimension = len(vectors[0].values)
+        # Estimate the metadata size by taking the average of 20 random vectors.
+        metadata_sizes = []
+        for _ in range(20):
+            metadata_sizes.append(len(json.dumps(random.choice(vectors).metadata)))
+        average_metadata_size = sum(metadata_sizes) / len(metadata_sizes)
+        # Estimate the size of the vector payload. 4 bytes per float.
+        vector_size = 4 * vector_dimension + average_metadata_size
+        # Cache the value.
+        self._upsert_batch_size = int(MAX_UPSERT_PAYLOAD_SIZE // vector_size)
+        return self._upsert_batch_size
+
+    @pinecone_observer.wrap({"type": "upsert_in_batches"})
+    async def upsert_in_batches(
+        self,
+        vectors: list[Vector],
+        batch_size: Optional[int] = None,
+        max_parallel_batches: int = 1,
+        batch_timeout: Optional[float] = None,
+    ) -> None:
+        """
+        Upsert vectors in batches.
+        Params:
+        - `vectors`: The vectors to upsert.
+        - `batch_size`: to control the number of vectors in each batch.
+        - `max_parallel_batches`: to control the number of batches sent concurrently.
+        - `batch_timeout`: to control the request timeout for each batch.
+        """
+        if batch_size is None:
+            batch_size = self._estimate_upsert_batch_size(vectors)
+
+        semaphore = asyncio.Semaphore(max_parallel_batches)
+
+        async def _upsert_batch(batch):
+            async with semaphore:
+                await self.upsert(vectors=batch, timeout=batch_timeout)
+
+        tasks = []
+        for batch in self._batchify(vectors, size=batch_size):
+            tasks.append(asyncio.create_task(_upsert_batch(batch)))
+
+        await asyncio.gather(*tasks)
+
+    @pinecone_observer.wrap({"type": "delete"})
+    async def delete(self, ids: list[str]) -> None:
+        """
+        Delete vectors by their ids.
+        Maximum number of ids in a single request is 1000.
```

Comment:
If there's a known limit, is it worth to validate before calling the API? Or batching?